data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/export/DataFileLoader.java

```diff
@@ -98,15 +113,19 @@ public void run() {
 
                     final long snapshotTime = progressState.getSnapshotTime();
                     final long eventVersionNumber = snapshotTime - VERSION_OVERLAP_TIME_FOR_EXPORT.toMillis();
-                    Record<Event> transformedRecord = new Record<>(
-                            recordConverter.convert(
-                                    record,
-                                    progressState.getSourceDatabase(),
-                                    progressState.getSourceTable(),
-                                    primaryKeys,
-                                    snapshotTime,
-                                    eventVersionNumber));
-                    bufferAccumulator.add(transformedRecord);
+                    final Event transformedEvent = recordConverter.convert(
+                            record,
+                            progressState.getSourceDatabase(),
+                            progressState.getSourceTable(),
+                            primaryKeys,
+                            snapshotTime,
+                            eventVersionNumber);
+
+                    if (acknowledgementSet != null) {
+                        acknowledgementSet.add(transformedEvent);
+                    }
+
+                    bufferAccumulator.add(new Record<>(transformedEvent));
```

Comment:
`BufferAccumulator` is not thread safe. I would recommend to instantiate this set with in the run method itself instead of passing to this thread. Or avoid using Accumulator if it is not really required.

Reply:
Good point! I moved the instantiation into the run method.